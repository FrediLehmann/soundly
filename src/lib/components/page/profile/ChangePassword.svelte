<script lang="ts" context="module">
  import { Input, Button, Flyin } from '$lib/components/atoms';
  import { AlertTriangle } from '$lib/Icons';
  import { supabase } from '$lib/supabase';
</script>

<script lang="ts">
  let flyin: Flyin;
  let editMode = false;
  let passwordError = '';
  let newPwd1 = '';
  let newPwd2 = '';

  const updatePwd = async () => {
    if (newPwd1 !== newPwd2) {
      passwordError =
        'Password verification failed. Please make sure the spelling of your password is correct.';
      return;
    }

    const { error } = await supabase.auth.update({ password: newPwd1 });
    if (error) {
      passwordError =
        'Something went wrong saving your new password. Please try again later.';
    } else {
      passwordError = '';
      newPwd1 = '';
      newPwd2 = '';
      editMode = false;
      flyin.show('Successfully changed password.', 'success');
    }
  };
</script>

<section>
  <h2>Password</h2>
  <p>
    Be sure to use a safe password to limit access to your account. Your
    password should contain at least 8 characters, containing lower- and
    uppercase letters, numbers and signs.
  </p>
  {#if editMode}
    <div class="input">
      <Input
        name="password"
        label="Password"
        type="password"
        required
        bind:value={newPwd1} />
    </div>
    <div class="input">
      <Input
        name="password"
        label="Repeat Password"
        type="password"
        required
        bind:value={newPwd2} />
    </div>
    {#if passwordError}
      <div class="warning">
        <div>
          <AlertTriangle width="1.75rem" height="1.75rem" />
        </div>
        <p>{passwordError}</p>
      </div>
    {/if}
    <Button on:click={() => (editMode = false)}>Cancel</Button>
    <Button style="primary" on:click={updatePwd}>Save new Password</Button>
  {:else}
    <Button on:click={() => (editMode = true)}>Change</Button>
  {/if}
</section>

<Flyin bind:this={flyin} />

<style>
  h2 {
    font-size: 1.2rem;
  }

  .input {
    max-width: 70%;
    margin-block-end: 1rem;
  }

  .warning {
    display: flex;
    align-items: center;
    gap: 1rem;
    background-color: hsla(var(--secondary-100));

    border: 1px solid hsla(var(--secondary-500));
    border-radius: 0.5rem;

    padding: 0.75rem 1rem;
    margin-block-end: 1rem;
  }

  .warning > p {
    margin: 0;
  }

  .warning > div {
    flex-shrink: 0;
  }

  section {
    max-width: 75%;
  }
</style>
