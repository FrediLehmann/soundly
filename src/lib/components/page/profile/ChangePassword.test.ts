/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom';
import { fireEvent, render } from '@testing-library/svelte';

import ChangePassword from './ChangePassword.svelte';

describe('<ChangePassword /> component', () => {
  test('renders correctly', () => {
    const { component } = render(ChangePassword);
    expect(component).toMatchSnapshot();
  });

  test('renders edit mode correctly', async () => {
    const { component, getByRole } = render(ChangePassword);
    await fireEvent.click(getByRole('button'));
    expect(component).toMatchSnapshot();
  });
});
