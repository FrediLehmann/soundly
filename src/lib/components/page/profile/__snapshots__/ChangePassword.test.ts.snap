// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChangePassword /> component renders correctly 1`] = `
ChangePassword {
  "$$": Object {
    "after_update": Array [],
    "before_update": Array [],
    "bound": Object {},
    "callbacks": Object {},
    "context": Map {},
    "ctx": Array [
      Flyin {
        "$$": Object {
          "after_update": Array [],
          "before_update": Array [],
          "bound": Object {},
          "callbacks": Object {},
          "context": Map {},
          "ctx": Array [
            false,
            undefined,
            undefined,
            [Function],
            [Function],
          ],
          "dirty": Array [
            -1,
          ],
          "fragment": Object {
            "c": [Function],
            "d": [Function],
            "i": [Function],
            "l": [Function],
            "m": [Function],
            "o": [Function],
            "p": [Function],
          },
          "not_equal": [Function],
          "on_destroy": Array [],
          "on_disconnect": Array [],
          "on_mount": Array [],
          "props": Object {
            "show": 4,
          },
          "root": <div>
            <section
              class="svelte-1ux9skm"
            >
              <h2
                class="svelte-1ux9skm"
              >
                Password
              </h2>
               
              <p>
                Be sure to use a safe password to limit access to your account. Your
    password should contain at least 8 characters, containing lower- and
    uppercase letters, numbers and signs.
              </p>
               
              <button
                class="svelte-1dsjpe6"
                data-type="secondary"
                type="button"
              >
                Change
              </button>
              
            </section>
             
            
          </div>,
          "skip_bound": false,
          "update": [Function],
        },
        "$capture_state": [Function],
        "$inject_state": [Function],
      },
      false,
      "",
      "",
      "",
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
    ],
    "dirty": Array [
      -1,
    ],
    "fragment": Object {
      "c": [Function],
      "d": [Function],
      "i": [Function],
      "l": [Function],
      "m": [Function],
      "o": [Function],
      "p": [Function],
    },
    "not_equal": [Function],
    "on_destroy": Array [
      [Function],
    ],
    "on_disconnect": Array [],
    "on_mount": Array [],
    "props": Object {},
    "root": <div>
      <section
        class="svelte-1ux9skm"
      >
        <h2
          class="svelte-1ux9skm"
        >
          Password
        </h2>
         
        <p>
          Be sure to use a safe password to limit access to your account. Your
    password should contain at least 8 characters, containing lower- and
    uppercase letters, numbers and signs.
        </p>
         
        <button
          class="svelte-1dsjpe6"
          data-type="secondary"
          type="button"
        >
          Change
        </button>
        
      </section>
       
      
    </div>,
    "skip_bound": false,
    "update": [Function],
  },
  "$capture_state": [Function],
  "$inject_state": [Function],
}
`;

exports[`<ChangePassword /> component renders edit mode correctly 1`] = `
ChangePassword {
  "$$": Object {
    "after_update": Array [],
    "before_update": Array [],
    "bound": Object {},
    "callbacks": Object {},
    "context": Map {},
    "ctx": Array [
      Flyin {
        "$$": Object {
          "after_update": Array [],
          "before_update": Array [],
          "bound": Object {},
          "callbacks": Object {},
          "context": Map {},
          "ctx": Array [
            false,
            undefined,
            undefined,
            [Function],
            [Function],
          ],
          "dirty": Array [
            -1,
          ],
          "fragment": Object {
            "c": [Function],
            "d": [Function],
            "i": [Function],
            "l": [Function],
            "m": [Function],
            "o": [Function],
            "p": [Function],
          },
          "not_equal": [Function],
          "on_destroy": Array [],
          "on_disconnect": Array [],
          "on_mount": Array [],
          "props": Object {
            "show": 4,
          },
          "root": <div>
            <section
              class="svelte-1ux9skm"
            >
              <h2
                class="svelte-1ux9skm"
              >
                Password
              </h2>
               
              <p>
                Be sure to use a safe password to limit access to your account. Your
    password should contain at least 8 characters, containing lower- and
    uppercase letters, numbers and signs.
              </p>
               
              <div
                class="input svelte-1ux9skm"
              >
                <div
                  class="container svelte-1ulq3bj"
                  data-required="true"
                  data-sr-only="false"
                >
                  <label
                    class="svelte-1ulq3bj"
                    for="password"
                  >
                    Password
                  </label>
                   
                  <input
                    class="svelte-1ulq3bj"
                    id="password"
                    required=""
                    type="password"
                  />
                   
                </div>
              </div>
               
              <div
                class="input svelte-1ux9skm"
              >
                <div
                  class="container svelte-1ulq3bj"
                  data-required="true"
                  data-sr-only="false"
                >
                  <label
                    class="svelte-1ulq3bj"
                    for="password"
                  >
                    Repeat Password
                  </label>
                   
                  <input
                    class="svelte-1ulq3bj"
                    id="password"
                    required=""
                    type="password"
                  />
                   
                </div>
              </div>
               
               
              <button
                class="svelte-1dsjpe6"
                data-type="secondary"
                type="button"
              >
                Cancel
              </button>
              
               
              <button
                class="svelte-1dsjpe6"
                data-type="primary"
                type="button"
              >
                Save new Password
              </button>
              
            </section>
             
            
          </div>,
          "skip_bound": false,
          "update": [Function],
        },
        "$capture_state": [Function],
        "$inject_state": [Function],
      },
      true,
      "",
      "",
      "",
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
      [Function],
    ],
    "dirty": Array [
      -1,
    ],
    "fragment": Object {
      "c": [Function],
      "d": [Function],
      "i": [Function],
      "l": [Function],
      "m": [Function],
      "o": [Function],
      "p": [Function],
    },
    "not_equal": [Function],
    "on_destroy": Array [
      [Function],
    ],
    "on_disconnect": Array [],
    "on_mount": Array [],
    "props": Object {},
    "root": <div>
      <section
        class="svelte-1ux9skm"
      >
        <h2
          class="svelte-1ux9skm"
        >
          Password
        </h2>
         
        <p>
          Be sure to use a safe password to limit access to your account. Your
    password should contain at least 8 characters, containing lower- and
    uppercase letters, numbers and signs.
        </p>
         
        <div
          class="input svelte-1ux9skm"
        >
          <div
            class="container svelte-1ulq3bj"
            data-required="true"
            data-sr-only="false"
          >
            <label
              class="svelte-1ulq3bj"
              for="password"
            >
              Password
            </label>
             
            <input
              class="svelte-1ulq3bj"
              id="password"
              required=""
              type="password"
            />
             
          </div>
        </div>
         
        <div
          class="input svelte-1ux9skm"
        >
          <div
            class="container svelte-1ulq3bj"
            data-required="true"
            data-sr-only="false"
          >
            <label
              class="svelte-1ulq3bj"
              for="password"
            >
              Repeat Password
            </label>
             
            <input
              class="svelte-1ulq3bj"
              id="password"
              required=""
              type="password"
            />
             
          </div>
        </div>
         
         
        <button
          class="svelte-1dsjpe6"
          data-type="secondary"
          type="button"
        >
          Cancel
        </button>
        
         
        <button
          class="svelte-1dsjpe6"
          data-type="primary"
          type="button"
        >
          Save new Password
        </button>
        
      </section>
       
      
    </div>,
    "skip_bound": false,
    "update": [Function],
  },
  "$capture_state": [Function],
  "$inject_state": [Function],
}
`;
